distribution/lib/Standard/Base/0.0.0-dev/src/Meta.enso

```diff
@@ -579,6 +588,72 @@ get_short_type_name typ = @Builtin_Method "Meta.get_short_type_name"
 get_constructor_declaring_type : Any -> Any
 get_constructor_declaring_type constructor = @Builtin_Method "Meta.get_constructor_declaring_type"
 
+instrumentor_builtin op args = @Builtin_Method "Meta.instrumentor_builtin"
+
+## PRIVATE
+   ADVANCED
+
+   Builder to create instrumentation for a function
+type Instrumentor
+    ## PRIVATE
+    Value impl
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed at the begining of node/expression
+       execution. The callback `fn` gets UUID of the node/expression that is
+       being executed and can return `Nothing` to continue regular execution
+       or anything else to skip the execution and just return given value.
+
+       Arguments:
+       - fn: The callback function accepting UUID.
+    on_enter self (fn : Text -> Any | Nothing) =
+        new = instrumentor_builtin "onEnter" [ self.impl, fn ]
+        Instrumentor.Value new
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed when a node/expression evaluation
+       is over. The callback `fn` gets UUID and the computed value. Usually
+       the value is _cached_ and returned from `on_enter` callback next time
+       the same expression is evaluated.
+
+       Arguments:
+       - fn: The callback function accepting UUID and computed value
+    on_return self (fn : Text -> Any -> Nothing) =
+        new = instrumentor_builtin "onReturn" [ self.impl, fn ]
+        Instrumentor.Value new
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed when a node/expression representing function is about to be called.
+       The callback `fn` shall accept three arguments. The UUID to identify the expression, the function to be
+       invoked and the arguments to pass to the function. The callback can return `Nothing`
+       (in such case the function gets executed with provided arguments) or some other value,
+       which is then returned instead of calling the function.
+
+       Arguments:
+       - fn: The callback function accepting UUID and function value
+    on_call self (fn : Text -> Function -> Vector Any -> Any | Nothing) =
+        new = instrumentor_builtin "onCall" [ self.impl, fn ]
+        Instrumentor.Value new
+
+    ## PRIVATE
+       ADVANCED
+
+       Activates configured instrumentor. Returns managed resource to
+       deactivate the instrumentor later.
+
+       Arguments:
+       - value: The value of the atom in the meta representation.
+    activate self =
+        finalize impl = instrumentor_builtin "deactivate" [ impl ]
+        managed = (instrumentor_builtin "activate" [ self.impl, finalize ]) : Managed_Resource
```

Comment:
I remember being told by @kustosz that keeping the `finalize` like this could mean it may capture parts of the related scope and then prevent actual cleanup.

I'm not sure if this is the case here and if it's still an issue, but I know we were usually using top-level functions for finalize.

I'd suggest to switch to a top-level function (essentially moving `finalize` to the module scope). Unless you have already tested that the resource is cleaned without trouble - in that case I guess it should be OK to stay as is.

Reply:
Indeed, the GC wasn't working. Here is a fix: 7d12a37 - now the Instrumentor can GC, but there is a danger of obtaining:
```
Execution finished with an error: com.oracle.truffle.api.instrumentation.UnwindException
        at <enso> acc<arg-2>(Internal)
        at <enso> acc<arg-2>(Instrumentor_Spec.enso:11:9-36)
        at <enso> acc(Instrumentor_Spec.enso:10-11)
        at <enso> repeat_gc(Instrumentor_Spec.enso:174:26-32)
        at <enso> Instrumentor_Spec.spec<arg-1>(Instrumentor_Spec.enso:182:13-44)
```
I need to find out how to prevent that exception.

Reply:
958ac95 - this seems to help. However releasing `Instrumentor` manually is certainly better approach than relying on a GC that may kick in at unexpected moments.